p4-16/psa/PSA.mdk

```diff
@@ -523,34 +523,71 @@ extern Checksum<W> {
 
 ## Counters
 
-Counters are a simple mechanism for keeping statistics about the
-packets that trigger a table in a Match Action unit.
+Counters are a mechanism for keeping statistics.  The control plane
+can read counter values.  A P4 program cannot read counter values,
+only update them.  If you wish to implement a feature involving
+sequence numbers, for example, use Registers instead (Section
+[#sec-registers]).
+
+Direct counters are counters associated with a particular P4 table,
+and are implemented by the extern `DirectCounter`.  There are also
+'indirect' counters, which are implemented by the extern `Counter`.
+The primary differences between direct counters and indirect counters
+are:
+
+- Number of independently updatable counter values:
+  - A single instantiation of a direct counter always contains as many
+    independent counter values as the number of entries in the table
+    with which it is associated.
+  - You must specify the number of independent counter values for an
+    indirect counter when instantiating it.  This number of counters
+    need not be the same as the size of any P4 table.
+- Where counter updates are allowed in the P4 program:
+  - For a direct counter, you may only invoke its `count` method from
+    inside the actions of the table with which it is associated, and
+    this always updates the counter value associated with the matching
+    table entry.
+  - For an indirect counter, you may invoke its `count` method
+    anywhere in the P4 program where extern object method invocations
+    are permitted (e.g. inside actions, or directly inside a control's
+    `apply` block), and every such invocation must specify the index
+    of the counter value to be updated.
```

Comment:
@vgurevich 

To address your concerns on portability: I agree with both your points -- PSA should define sufficient functionality to be useful and that can be ported to different targets. And that we should strive to hide the target from the control plane, such that not only P4 programs are portable, but as much of the control plane is portable as well.

As you know, I have been pushing for getting the control plane API in the PSA spec so that one can look at the PSA spec and know how to use these externs without leafing through several target reference manuals. 
However, I do not think we are ready to change the language to support the expression of control plane APIs in P4, and certainly not in the timeframe we are discussing to push out a first draft of the PSA. Given these constraints, I think we should have at least a textual definition of these APIs. I'm sure it is not fully satisfactory, but it is a reasonable compromise toward getting people's mindshare.